cpp/tests/datetime/datetime_ops_test.cu

```diff
@@ -0,0 +1,154 @@
+
+  using namespace cudf::test;
+  using namespace simt::std::chrono;
+  using Rep = typename TypeParam::rep;
+  using Period = typename TypeParam::period;
+
+  auto start = milliseconds(-2500000000000); // Sat, 11 Oct 1890 19:33:20 GMT
+  auto stop_ = milliseconds( 2500000000000); // Mon, 22 Mar 2049 04:26:40 GMT
+  auto test_timestamps = generate_timestamps<Rep, Period>(this->size(),
+                                                          time_point_ms(start),
+                                                          time_point_ms(stop_));
+
+  auto timestamp_col = cudf::make_timestamp_column(this->type(), this->size(),
+                                                   cudf::mask_state::UNALLOCATED,
+                                                   this->stream(), this->mr());
+
+  cudf::mutable_column_view timestamp_view = *timestamp_col;
+
+  CUDA_TRY(cudaMemcpy(timestamp_view.data<Rep>(),
+    thrust::raw_pointer_cast(test_timestamps.data()),
+    test_timestamps.size() * sizeof(Rep), cudaMemcpyDefault));
+
+  auto expected_years = fixed_width_column_wrapper<int16_t>{1890, 1906, 1922, 1938, 1954, 1970, 1985, 2001, 2017, 2033};
+  auto expected_months = fixed_width_column_wrapper<int16_t>{10, 8, 6, 4, 2, 1, 11, 9, 7, 5};
+  auto expected_days = fixed_width_column_wrapper<int16_t>{11, 16, 20, 24, 26, 1, 5, 9, 14, 18};
+  auto expected_weekdays = fixed_width_column_wrapper<int16_t>{6, 4, 2, 7, 5, 4, 2, 7, 5, 3};
+  auto expected_hours = fixed_width_column_wrapper<int16_t>{19, 20, 21, 22, 23, 0, 0, 1, 2, 3};
+  auto expected_minutes = fixed_width_column_wrapper<int16_t>{33, 26, 20, 13, 6, 0, 53, 46, 40, 33};
+  auto expected_seconds = fixed_width_column_wrapper<int16_t>{20, 40, 0, 20, 40, 0, 20, 40, 0, 20};
+
+  // Special cases for timestamp_D: zero out the hh/mm/ss cols and +1 the expected weekdays
+  if (std::is_same<TypeParam, cudf::timestamp_D>::value) {
+    expected_days = fixed_width_column_wrapper<int16_t>{12, 17, 21, 25, 27, 1, 5, 9, 14, 18};
+    expected_weekdays = fixed_width_column_wrapper<int16_t>{7, 5, 3, 1, 6, 4, 2, 7, 5, 3};
+    expected_hours = fixed_width_column_wrapper<int16_t>{0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+    expected_minutes = fixed_width_column_wrapper<int16_t>{0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+    expected_seconds = fixed_width_column_wrapper<int16_t>{0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+  }
+
+  expect_columns_equal(*cudf::datetime::extract_year(timestamp_view), expected_years);
+  expect_columns_equal(*cudf::datetime::extract_month(timestamp_view), expected_months);
+  expect_columns_equal(*cudf::datetime::extract_day(timestamp_view), expected_days);
+  expect_columns_equal(*cudf::datetime::extract_weekday(timestamp_view), expected_weekdays);
+  expect_columns_equal(*cudf::datetime::extract_hour(timestamp_view), expected_hours);
+  expect_columns_equal(*cudf::datetime::extract_minute(timestamp_view), expected_minutes);
+  expect_columns_equal(*cudf::datetime::extract_second(timestamp_view), expected_seconds);
+}
```

Comment:
I still need to add tests for columns with null values.

Reply:
Also test that empty columns produce empty results (but no errors), and test that exceptions are thrown if non-datetime columns are passed.